File: agents/triple-storages/rastore/src/main/scala/ru/magnetosoft/rastore/server/MessageParser.scala
package ru.magnetosoft.rastore.server

import scala.collection.mutable.Set
import scala.collection.mutable.HashMap

import ru.magnetosoft.rastore.util.Common
import ru.magnetosoft.rastore.core.Triplet
import ru.magnetosoft.rastore.core.TripletModifier
import ru.magnetosoft.rastore.core.Logger

import scala.collection.jcl.ArrayList

object MessageParser {

  val logger = new Logger(MessageParser)

  def functionsFromMessage(message: String): Set[OntoFunction] = {

    logger.debug("Got message for parsing [ " + message + " ]")

    val functions = HashMap[String, OntoFunction]()
    val argumentsMap = HashMap[String, Set[Triplet]]()
    
    val lines = split(message)

    for(line <- lines) {
      val triplet = Common.tripletFromLine(line)
      
      if (triplet.subj == "subject") {
        if (triplet.mod == TripletModifier.TripletSet.id) { // триплет с набором триплетов в объекте ?
          val new_triplet_uid = "uid-" + triplet.obj.hashCode
          for(s <- split(triplet.obj)) {
            if (argumentsMap.get(new_triplet_uid) == None)
              argumentsMap += (new_triplet_uid -> Set[Triplet]())
            argumentsMap(new_triplet_uid) += new Triplet(0, new_triplet_uid, "argument", s, TripletModifier.TripletSet.id)
          }

          if (functions.get(new_triplet_uid) == None) {
            triplet.obj_=(new_triplet_uid)
              triplet.mod_=(TripletModifier.Literal.id)
                functions += (new_triplet_uid -> new OntoFunction(triplet, Set[Triplet]()))
          }

        } else if (functions.get(triplet.obj) == None)
          functions += (triplet.obj -> new OntoFunction(triplet, Set[Triplet]()))
      } else if (triplet.pred == "argument") {
        if (argumentsMap.get(triplet.subj) == None)
          argumentsMap += (triplet.subj -> Set[Triplet]())
        argumentsMap(triplet.subj) += triplet
      }
    }

    val result = Set[OntoFunction]()
    for(key <- argumentsMap.keySet) {
      functions(key).arguments ++ argumentsMap(key)
      result += functions(key)
    }

    logger.debug("OntoFunctions created [ " + result + " ]")
    return result
  }

  def split(tripletsLine: String): Set[String] = {

    val result = Set[String]()
    val indexes = new ArrayList[Int]()

    var isBeetweenTokens = false
    var delim_number = 1

    var prev_delimiter = ' '

    def searchParams(char: Char, delim_number: Int) = {
      prev_delimiter = char
      delim_number match {
        case 0 => (false, 1)
        case 1 => (true, 2)
        case 2 => (false, 3)
        case 3 => (true, 4)
        case 4 => (false, 5)
        case 5 => (true, 6)
        case 6 => (true, 0)
      }
    }
    for(i <- 0 until tripletsLine.length) {
      val c = tripletsLine.charAt(i)
      if (c != ' ') { // если пробел, пох
        val isProcessNeeded = 
          delim_number match {
            case 0 => c == '<'
            case 1 => c == '>'
            case 2 => c == '<'
            case 3 => c == '>'
            case 4 => {
              (c == '"' || c == '<' || c == '{')
            }
            case 5 => 
              prev_delimiter match {
                case '"' => (c == '"')
                case '<' => (c == '>')
                case '{' => (c == '}')
                case _ => false
              }
            case 6 => c == '.'
            case _ => false
          }
        if (isProcessNeeded) { // если один из разделителей, то не пох
          if (c == '.') // ага, конец триплета
            indexes += i // закидываем индекс конца триплета в список
          val search_params = searchParams(c, delim_number) // берем следующий tuple (разделитель, мы_между_тоекнами, номер_разделителя) для поиска
          //          nextChar = search_params._1
          isBeetweenTokens = search_params._1
          delim_number = search_params._2
        } else if (isBeetweenTokens) {
          val search_params = if (delim_number == 0 && indexes.size > 0) {
            indexes.remove(indexes.size - 1)
            searchParams(prev_delimiter, 5)
          } else searchParams(c, delim_number)
          //          nextChar = search_params._1
          isBeetweenTokens = search_params._1
          delim_number = search_params._2
        }
      }
    }
    
    var prev = 0
    for(i <- indexes) {
      result += tripletsLine.substring(prev, i + 1).trim
      prev = i + 1
    }
    result
  }
}

File: agents/magnetico-authorization/autotest/java/src/ru/magnetosoft/magnet/messaging/MessageParser.java
/**
 * Copyright (c) 2006-2009, Magnetosoft, LLC
 * All rights reserved.
 *
 * Licensed under the Magnetosoft License. You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.magnetosoft.ru/LICENSE
 * 
 * file: MessageParser.java
 */

package ru.magnetosoft.magnet.messaging;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class MessageParser {
    
    private boolean sp1;
    private int sp2;
    private char prev_delimiter = ' ';
    
    public MessageParser() {
    }
    
    public List<OntoFunction> functionsFromMessage(String message) {
	
	//	System.out.println("Получил сообщение для парсинга [ " + message + " ]");
	
	HashMap<String, OntoFunction> functions = new HashMap<String, OntoFunction>();
	HashMap<String, List<Triple>> argumentsMap = new HashMap<String, List<Triple>>();
	
	List<String> lines = split(message);
	
	for (String line : lines) {
	    
	    Triple triple = TripleUtils.getTripleFromLine(line);
	    
	    if (triple.getPred().equals(Predicates.SUBJECT)) { 
		if (triple.getMod() == TripleModifier.TripleSet.ordinal()) {
		    
		    for (String s : split(triple.getObj())) {
			if (argumentsMap.get(triple.getSubj()) == null) {
			    argumentsMap.put(triple.getSubj(), new ArrayList<Triple>());
			}
			argumentsMap.get(triple.getSubj()).add(new Triple(0, triple.getSubj(), Predicates.FUNCTION_ARGUMENT, s,
									  TripleModifier.TripleSet.ordinal()));
		    }
		    
		    if (functions.get(triple.getSubj()) == null) {
			functions.put(triple.getSubj(), new OntoFunction(triple, new ArrayList<Triple>()));
		    }
		    
		} else if (functions.get(triple.getSubj()) == null)
		    functions.put(triple.getSubj(), new OntoFunction(triple, new ArrayList<Triple>()));
	    } else if (triple.getPred().equals(Predicates.FUNCTION_ARGUMENT)) {
		if (argumentsMap.get(triple.getSubj()) == null) {
		    argumentsMap.put(triple.getSubj(), new ArrayList<Triple>());
		}
		argumentsMap.get(triple.getSubj()).add(triple);
	    } else if (triple.getPred().equals(Predicates.REPLY_TO)) {
		if (functions.get(triple.getSubj()) == null) {
		    functions.put(triple.getSubj(), new OntoFunction(triple, new ArrayList<Triple>()));
		}
		functions.get(triple.getSubj()).setReplyTo(triple.getObj());
	    }
	}
	
	ArrayList<OntoFunction> result = new ArrayList<OntoFunction>();
	
	for (String key : argumentsMap.keySet()) {
	    functions.get(key).getArguments().addAll(argumentsMap.get(key));
	    result.add(functions.get(key));
	}
	
	//	System.out.println("Созданы объекты OntoFunction [ " + result.toString() + " ]");
	return result;
    }

    public Triple getArgumentByPredicate(List<Triple> args, String argName) {
	Triple result = null;
	for (Triple triple : args) {
	    if (triple.getPred().equals(argName)) {
		return triple;
	    }
	}
	return result;
    }
    
    public List<String> split(String tripletsLine) {
	List<String> result = new ArrayList<String>();
	
	if (tripletsLine != null && tripletsLine.trim().length() > 0) {
	    ArrayList<Integer> indexes = new ArrayList<Integer>();
	    
	    boolean isBeetweenTokens = false;
	    int delim_number = 1;
	    boolean isProcessNeeded;
	    
	    for (int i = 0; i < tripletsLine.length(); i++) {
		char c = tripletsLine.charAt(i);
		if (c != ' ') { // если пробел, пох
		    switch (delim_number) {
		    case 0:
			isProcessNeeded = c == '<';
			break;
		    case 1:
			isProcessNeeded = c == '>';
			break;
		    case 2:
			isProcessNeeded = c == '<';
			break;
		    case 3:
			isProcessNeeded = c == '>';
			break;
		    case 4:
			isProcessNeeded = (c == '"' || c == '<' || c == '{');
			break;
		    case 5:
			switch (prev_delimiter) {
			case '"':
			    isProcessNeeded = (c == '"');
			    break;
			case '<':
			    isProcessNeeded = (c == '>');
			    break;
			case '{':
			    isProcessNeeded = (c == '}');
			    break;
			default:
			    isProcessNeeded = false;
			}
			break;
		    case 6:
			isProcessNeeded = c == '.';
			break;
		    default:
			isProcessNeeded = false;
		    }
		    if (isProcessNeeded) { // если один из разделителей, то не пох
			if (c == '.') // ага, конец триплета
			    indexes.add(i); // закидываем индекс конца триплета в
			// список
			searchParams(c, delim_number); // берем следующий tuple
			// (разделитель,
			// мы_между_тоекнами,
			// номер_разделителя) для
			// поиска
			// nextChar = search_params._1
			isBeetweenTokens = sp1;
			delim_number = sp2;
		    } else if (isBeetweenTokens) {
			if (delim_number == 0 && indexes.size() > 0) {
			    indexes.remove(indexes.size() - 1);
			    searchParams(prev_delimiter, 5);
			} else
			    searchParams(c, delim_number);
			// nextChar = search_params._1
			isBeetweenTokens = sp1;
			delim_number = sp2;
		    }
		}
	    }
	    
	    int prev = 0;
	    for (Integer i : indexes) {
		result.add(tripletsLine.substring(prev, i + 1).trim());
		prev = i + 1;
	    }
	}
	//	System.out.println(String.format("Получены токены триплетов из строки [ %s ]", result));
	return result;
    }
    
    public void searchParams(char ch, int delim_number) {
	prev_delimiter = ch;
	switch (delim_number) {
	case 0:
	    sp1 = false;
	    sp2 = 1;
	    break;
	case 1:
	    sp1 = true;
	    sp2 = 2;
	    break;
	case 2:
	    sp1 = false;
	    sp2 = 3;
	    break;
	case 3:
	    sp1 = true;
	    sp2 = 4;
	    break;
	case 4:
	    sp1 = false;
	    sp2 = 5;
	    break;
	case 5:
	    sp1 = true;
	    sp2 = 6;
	    break;
	case 6:
	    sp1 = true;
	    sp2 = 0;
	    break;
	}
    }
    
}